### PR TITLE
Productivity measurement

### PR DESCRIPTION
**Pull request by angelala3252**
This PR responds to [issue #173](https://github.com/MLDSAI/OpenAdapt/issues/173) by gathering basic productivity information from the most recent recording and displaying it on an HTML page (similar to `visualize.py`). The screenshots on the left are those where a window event occurs (changing tabs, changing windows, etc.), and there is some basic information displayed to the right of that screenshot about what was done on that window (# of clicks, # of keys...).

Ex:
Overall info at the top
image
1st window
image
2nd window
image

Instructions to run:
```
git remote add angelala3252 https://github.com/angelala3252/OpenAdapt.git
git fetch angelala3252
git checkout feat/productivity
python -m openadapt.productivity 
```
For a better example try to switch tabs and do multiple actions on each tab, and try pausing for a bit (you can change the number of seconds that gets recognized as a long pause at `productivity.MAX_GAP_SECONDS`.

### CODE REVIEW

**Comment by angelala3252** [comment]
Updates:

Added some info to the overview section
image
And added time spent on each window (although the window-by-window section has some issues and isn't 100% accurate)
image

Now working on identifying repetitive tasks from the `ActionEvents` and making it more accurate. I've tried using https://github.com/oemof/cydets but it identifies almost everything as a cycle, even if it doesn't repeat and has no pattern to it. I've tried a brute force method but it was way too slow. I will continue trying to optimize my brute force method, as well as researching to see if there are any algorithms I can use.

**Comment by angelala3252** [comment]
I've implemented Brent's algorithm to try to find cycles in the ActionEvents. I tested it on recordings containing repetitions of the simple task of 1. click start 2. click word 3. click start 4. click settings. In a recording containing two back-to-back repetitions of the simple task, it found 2 repetitions. In another recording containing two repetitions of the simple task separated by an unrelated click, it also found 2 repetitions. I will have to do more testing on different examples, but I think this is a great result so far.

**Comment by abrichr** [comment]
@angelala3252 any thoughts on https://huggingface.co/blog/autoformer ?

> Autoformer builds upon the traditional method of decomposing time series into seasonality and trend-cycle components. This is achieved through the incorporation of a Decomposition Layer, which enhances the model's ability to capture these components accurately. Moreover, Autoformer introduces an innovative auto-correlation mechanism that replaces the standard self-attention used in the vanilla transformer. This mechanism enables the model to utilize period-based dependencies in the attention, thus improving the overall performance.

**Comment by abrichr** [comment]
@angelala3252 another way to think about this would be to measure the "similarity" between sequences, rather than identifying equality.

e.g. https://arxiv.org/abs/2306.05426

**Comment by angelala3252** [comment]
Using the longest repeated non-overlapping substring algorithm found [here](https://www.geeksforgeeks.org/longest-repeating-and-non-overlapping-substring/), I am now able to identify any task that repeats in a recording, whether the tasks are back to back or not. 🥳

**Comment by angelala3252** [comment]
Currently, the task identification has inconsistent/unpredictable behaviour on longer recordings (the identified task contains the current task and part of the next task), so I'm working on trying to fix this issue.

**Comment by angelala3252** [comment]
Example pdf: 
[Productivity metrics for recording-28.pdf](https://github.com/OpenAdaptAI/OpenAdapt/files/12264540/Productivity.metrics.for.recording-28.pdf)

File: openadapt/productivity.py [hunk 1]
```diff
@@ -0,0 +1,674 @@
...
+
+    $ python openadapt/productivity.py"
+"""
+
+from pprint import pformat
+from threading import Timer
+import os
+import string
+from typing import Optional, Tuple
+
```
Comment (KrishPatel13) [inline openadapt/productivity.py:15]: The order of imports is incorrect here.

Please see Contributing md file for details

**Review by KrishPatel13 — COMMENTED** [review]
@angelala3252 I would strong encourage you to run:

`black <on all of the changed python files>`
`flake8 <on all of the changed python files>`
and then fix all errors if any

Optinoal but I do it:
`pylint <on all of the changed python files>`

File: openadapt/productivity.py (hunk 1 above)
Comment (KrishPatel13) [reply openadapt/productivity.py:15]: Other flake8 issues:
https://github.com/OpenAdaptAI/OpenAdapt/pull/474#issuecomment-1692673244

**Comment by angelala3252** [comment]
@KrishPatel13 I believe I finished this PR before I installed the pre-commit which is why I missed some flake8 errors previously 😅  This new PR should resolve the errors
https://github.com/OpenAdaptAI/OpenAdapt/pull/480

**Comment by abrichr** [comment]
Thank you for the great work @angelala3252 !

Regarding your comment on Aug 4:

> Example pdf:
> [Productivity metrics for recording-28.pdf](https://github.com/OpenAdaptAI/OpenAdapt/files/12264540/Productivity.metrics.for.recording-28.pdf)

This appears to have the productivity metrics cut off or otherwise ineligible:

image

Would be great to polish this off so we can show it off on the README 😄

**Pull request merged**